File: qetah/public/modules/downloads.js
function price(data) {
  if (data.price) {
    return `$<span>${data.price}</span>`;
  } else {
    return `$<span>${data.discount}</span><del>${data.price}<del>`;
  }
}

function card(item) {
  return `
    <article class="media media-product">
    <a class="media-image" href="/product/${item.product_name}/">
        <img src="${item.image}" class="image" alt="${item.alt}" />
    </a>
    <div class="media-body">
        <a href="/shop/${item.category_name}/" class="media-badge">
          ${ item.category_title }
        </a>
        <h3 class="media-title text-ellipsis">
            <a href="/product/${item.product_name}/">
              ${item.product_title} (${item.variation_title}) 
            </a>
        </h3>
        <p class="media-desc">${price(item)}</p>
        <a href="${item.file}" download="${item.file}" class="btn btn-pill btn-sm btn-primary">
            Download<i class="bx bx-download icon-end"></i>
        </a>
    </div>
  </article>`;
}

function listing(page = 1, append = true) {
  let count = document?.querySelector("#count");
  let output = document.querySelector("#listing");
  let loadmore = document.querySelector("#loadmore");
  let sort = document.querySelector("#sort").value;
  let template = "";
  let query = searchParams({
    limit: 6,
    page: page,
    sort: sort,
  });

  fetch(ApiUrl("qetah.events.downloads.listing/?" + query))
    .then((res) => res.json())
    .then((data) => {
      console.log(data);
      data = data.message;
      if (data.count == 0) {
        count ? (count.innerHTML = "(0)") : null;
        output.innerHTML = "<p class='no-records'>No products exists!</p>";
      } else if (data.count != 0) {
        count ? (count.innerHTML = "(" + data.count + ")") : null;
        data.results.forEach((item) => { template += card(item);});
        append ? (output.innerHTML += template) : (output.innerHTML = template);
      }
      if (loadmore) {
        if (data.next) {
          loadmore.value = data.next;
          loadmore.disabled = false;
          loadmore.style.display = "inline-block";
        } else {
          loadmore.value = 0;
          loadmore.disabled = true;
          loadmore.style.display = "none";
        }
      }
    })
    .finally(() => {});
}

listing();

on("change", "#sort", function (e) {
  e.preventDefault();
  listing(1, false);
});

on("submit", "#searchForm", function (e) {
  e.preventDefault();
  listing(1, false);
});

on("click", "#loadmore", function (e) {
  e.preventDefault();
  listing(e.target.value, true);
});

on("change", "[name='category']", function (e) {
  e.preventDefault();
  var values = [];
  document.querySelectorAll(".category").forEach((el) => {
    if (el.checked) {
      values.push(el.value);
    }
  });
  document.querySelector("#categories").value = values;
  listing(1, false);
});

File: qetah/events/downloads.py
import frappe


@frappe.whitelist()
def listing(**data):
    limit = data.get("limit", 6)
    page = data.get("page", 1)
    limit = int(limit)
    page = int(page)
    user = frappe.session.user
    sort = data.get("sort", "desc")

    count = frappe.db.count(
        "Download",
        filters={"owner": user},
    )

    start = (page - 1) * limit
    end = start + limit

    query = f"""
        SELECT d.name AS download_name, v.price, v.discount, p.alt, 
        p.featured_image, p.name AS product_name, p.title AS product_title,
        v.name AS variation_name, v.title AS variation_title, p.file,
        c.name AS category_name, c.title AS category_title
        FROM `tabDownload` AS d
        LEFT JOIN `tabProduct Variation` AS v ON v.name = d.variation
        LEFT JOIN `tabProduct` AS p ON p.name = v.parent
        LEFT JOIN `tabProduct Category` AS c ON c.name = p.category
        WHERE d.owner = %(user)s ORDER BY d.creation {sort}
    """

    products = frappe.db.sql(query, {"user": user}, as_dict=True)
    print(products)
    results = []
    for product in products:
        results.append(
            {
                "download": product.get("download"),
                "product_name": product.get("product_name"),
                "product_title": product.get("product_title"),
                "variation_name": product.get("variation_name"),
                "variation_title": product.get("variation_title"),
                "category_name": product.get("category_name"),
                "category_title": product.get("category_title"),
                "image": product.get("featured_image"),
                "alt": product.get("alt"),
                "file": product.get("file"),
                "price": product.get("price"),
                "discount": product.get("discount"),
            }
        )

    next_page = None if end >= count else page + 1
    back_page = None if page <= 1 else page - 1

    return {
        "count": count,
        "next": next_page,
        "back": back_page,
        "results": results,
    }
